### ISSUE TITLE
Consider staging -> publish flow

### ISSUE THREAD

**Issue by ravwojdyla** [related-sciences/public-data-source#5]
I believe current flow handle the "happy path" well, but in case of failure it might lead to partial/incomplete results - say if there is a failure in the middle of `fs.upload`. Usually this is handled by upload to temp directory and then final move operation + marker file to make this process more deterministic. We could have a util function to handle this kind of upload.

**Comment by eric-czech** [comment]
Makes sense, I think that would be a great addition.

**Comment by ravwojdyla** [comment]
@eric-czech regarding efficiency - which part are you worried about?

**Comment by eric-czech** [comment]
I thought moving folders in GCS meant that it would copy all the objects and then delete the original rather than being an atomic operation like on filesystems.  I remember it taking forever with `gsutil mv` in the past although Jeff just did this recently with some UKB data within a GCE VM and it was very fast for a large dataset.  Maybe my perceived slowness is more related to `gsutil` or working outside of GCP.

**Comment by ravwojdyla** [comment]
@eric-czech I *believe* that should not be an issues for reasonable number of files (at least it has not been in my experience). There might be an issues if you have many Ks of files to copy and a single thread to do the move, but in most cases you should not have that issue. Another way to solve this problem is to have for example a hash/uuid in the output path and some way to register successful outputs - but this requires a database to keep tract of successful outputs (this could be part of a catalog service, if there was one).

**Comment by eric-czech** [comment]
Makes sense.  I don't know how it's managed internally with fsspec but I'm sure it's nothing fancy.

Whether or not we want to use fsspec for this is certainly another good question.  I went that route since dask uses it and it's cross-platform but I would imagine there may be clients that already have some staging logic built in for simulated atomic uploads.  Do you know of any that do that off-hand?

**Comment by ravwojdyla** [comment]
@eric-czech actually - you would think so, but I don't know any FS abstractions that would have that built in, but to be fair it's relatively simple to implement.
